app/services/foreman_rh_cloud/ui_request_forwarder.rb

```diff
@@ -0,0 +1,106 @@
+require 'rest-client'
+
+module ForemanRhCloud
+  class UIRequestForwarder
+    include ForemanRhCloud::GatewayRequest
+
+    def forward_request(original_request, path, controller_name, user, organization, location)
+      TagsAuth.new(user, logger).update_tag if scope_request?(original_request)
+
+      forward_params = prepare_forward_params(original_request, user: user, organization: organization, location: location).to_a
+      logger.debug("Request parameters for UI request: #{forward_params}")
+
+      forward_payload = prepare_forward_payload(original_request, controller_name)
+
+      logger.debug("User agent for UI is: #{http_user_agent(original_request)}")
+
+      request_opts = prepare_request_opts(original_request, path, forward_payload, forward_params)
+
+      logger.debug("Sending request to: #{request_opts[:url]}")
+
+      execute_cloud_request(request_opts)
+    end
+
+    def prepare_tags(user, organization, location)
+      [
+        CGI.escape(TagsAuth.auth_tag_for(user)),
+        CGI.escape("satellite/organization=#{organization}"),
+        CGI.escape("satellite/location=#{location}"),
+      ].map { |tag_value| [:tag, tag_value] }
```

Comment:
Some APIs require `tags=` instead of `tag=`. Unfortunately this is not consistent.

Reply:
So do we need a list of API endpoints that accept `tags` instead of `tag`. 
What do you think about exposing it as [settings](https://github.com/RedHatInsights/iop-gateway/blob/d581ba24dd5722ee7011b9f285de0e72c619e8ff/nginx/common.d/60-smart-proxy.conf#L13) metadata by the proxy? It looks like an implementation detail that Foreman shouldn't be aware of.

Reply:
Ideally it could be read from openapi spec. This is implementation detail of the adapter.

Reply:
Each services exposes its own spec. Here is a list of common ones:
* Inventory https://console.redhat.com/api/inventory/v1/openapi.json
* Advisor https://console.redhat.com/api/insights/v1/openapi.json
* Vuln https://console.redhat.com/api/vulnerability/v1/openapi.json

The paths should be available via iop gateway. The plugin can fetch and cache them.

All APIs can be found at https://console.redhat.com/docs/api